python/pyarrow/tests/test_gpu.py

```diff
@@ -0,0 +1,578 @@
+    writer = pa.CudaBufferWriter(cbuf)
+    arr, buf = make_random_buffer(size=size, target='host')
+
+    assert writer.buffer_size == 0
+    writer.set_buffer_size(100)
+    assert writer.buffer_size == 100
+
+    writer.write(buf.slice(length=0))
+    assert writer.tell() == 0
+
+    writer.write(buf.slice(length=10))
+    writer.set_buffer_size(200)
+    assert writer.buffer_size == 200
+    assert writer.num_bytes_buffered == 0
+
+    writer.write(buf.slice(offset=10, length=300))
+    assert writer.num_bytes_buffered == 0
+
+    writer.write(buf.slice(offset=310, length=200))
+    assert writer.num_bytes_buffered == 0
+
+    writer.write(buf.slice(offset=510, length=390))
+    writer.write(buf.slice(offset=900, length=100))
+
+    writer.flush()
+
+    buf2 = cbuf.copy_to_host()
+    assert buf2.size == size
+    arr2 = np.frombuffer(buf2, dtype=np.uint8)
+    assert arr2.size == size
+    assert (arr2 == arr).all()
+
+
+@gpu_support
+def test_CudaBufferReader():
+
+    size = 1000
+    arr, cbuf = make_random_buffer(size=size, target='device')
+
+    reader = pa.CudaBufferReader(cbuf)
+    reader.seek(950)
+    assert reader.tell() == 950
+
+    data = reader.read(100)
+    assert len(data) == 50
+    assert reader.tell() == 1000
+
+    reader.seek(925)
+    arr2 = np.zeros(100, dtype=np.uint8)
+    n = reader.readinto(arr2)
+    assert n == 75
+    assert reader.tell() == 1000
+    assert (arr2[:75] == arr[925:]).all()
+
+    reader.seek(0)
+    assert reader.tell() == 0
+    buf2 = reader.read_buffer()
+    arr2 = np.frombuffer(buf2.copy_to_host(), dtype=np.uint8)
+    assert (arr2 == arr).all()
+
+
+def make_recordbatch(length):
+    schema = pa.schema([pa.field('f0', pa.int16()),
+                        pa.field('f1', pa.int16())])
+    a0 = pa.array(np.random.randint(0, 255, size=length, dtype=np.int16))
+    a1 = pa.array(np.random.randint(0, 255, size=length, dtype=np.int16))
+    batch = pa.RecordBatch.from_arrays([a0, a1], schema)
+    return batch
+
+
+@gpu_support
+def test_batch_serialize():
+    batch = make_recordbatch(10)
+    hbuf = batch.serialize()
+    cbuf = pa.cuda_serialize_record_batch(batch, global_context)
+    pa.cuda_read_record_batch(batch.schema, cbuf)
+    buf = cbuf.copy_to_host()
+    assert hbuf.equals(buf)
+    batch2 = pa.read_record_batch(buf, batch.schema)
+    assert hbuf.equals(batch2.serialize())
+    assert batch.num_columns == batch2.num_columns
+    assert batch.num_rows == batch2.num_rows
+    assert batch.column(0).equals(batch2.column(0))
+    assert batch.equals(batch2)
+
+
+@gpu_ipc_support
+@pytest.mark.skip(reason='open_ipc_buffer fails with unknown reason')
+def test_IPC():
+    size = 1000
+    arr, cbuf = make_random_buffer(size=size, target='device')
+    ipc_handle = cbuf.export_for_ipc()
+    b = ipc_handle.serialize()
+    ipc_handle2 = pa.CudaIpcMemHandle.from_buffer(b)
+    # cuIpcOpenMemHandle fails here with code 201:
+    ipc_buf = cbuf.context.open_ipc_buffer(ipc_handle2)
```

Comment:
I think you have to do this in a different process for it to work. See the Ruby tests for this https://github.com/apache/arrow/blob/master/c_glib/test/test-gpu-cuda.rb#L49

Reply:
Yes, I got test_IPC working using `multiprocessing.Process` with spawn method.